tempesta_fw/addr.c

```diff
@@ -400,44 +384,44 @@ EXPORT_SYMBOL(tfw_addr_eq);
 int
 tfw_addr_ifmatch(const TfwAddr *server, const TfwAddr *listener)
 {
-	switch (server->family) {
-	case AF_INET:
-		do {
-			unsigned laddr = listener->v4.sin_addr.s_addr;
-			unsigned saddr = server->v4.sin_addr.s_addr;
-			if (listener->v4.sin_port != server->v4.sin_port)
-				return 0;
-			if (laddr == 0) {
-				if (IN_LOOPBACK(ntohl(saddr)))
-					return 1;
-				/* TODO: check if client addr is
-				* one of interface adresses */
-			}
-		} while (0);
-		return tfw_addr_eq_inet(&server->v4, &listener->v4);
-		break;
-	case AF_INET6:
-		if (listener->v6.sin6_port != server->v6.sin6_port)
+	if (unlikely(server->sin6_family != AF_INET6)) {
+		TFW_ERR("Unexpected protocol family\n");
+		BUG();
+		return 0;
+	}
+
+	if (tfw_addr_is_v4mapped(server) && tfw_addr_is_v4mapped(listener)) {
+		__be32 laddr = tfw_addr_v4addr(listener);
+		__be32 saddr = tfw_addr_v4addr(server);
+
+		if (tfw_addr_port(listener) != tfw_addr_port(server))
 			return 0;
-		if (!(listener->v6.sin6_addr.in6_u.u6_addr32[0] |
-			listener->v6.sin6_addr.in6_u.u6_addr32[1] |
-			listener->v6.sin6_addr.in6_u.u6_addr32[2] |
-			listener->v6.sin6_addr.in6_u.u6_addr32[3]))
-		{
-			/* listener = [::] */
-			if (IN6_LOOPBACK(server->v6.sin6_addr))
-			{
-				/* backend = [::1] */
+
+		if (laddr == 0) {
+			if (IN_LOOPBACK(ntohl(saddr)))
 				return 1;
-			}
-			/* TODO: same as in v4 case */
+				/* TODO: check if client addr is
+				 * one of interface adresses
+				 */
 		}
-		return tfw_addr_eq_inet6(&server->v6, &listener->v6);
-	default:
-		TFW_ERR("Unknown protocol family\n");
-		BUG();
+
+		return tfw_addr_eq(server, listener);
 	}
-	return 0;
+
+	if (tfw_addr_port(listener) != tfw_addr_port(server))
```

Comment:
It seems, that the same check with exit is repeated here and in the `v4mapped` block above. Perhaps it is better to leave only one this check - before the `v4` verification.

Reply:
Moved the check to the beginning of the function.